### PR TITLE
Implement parallel::inplace_merge.

### PR DESCRIPTION
**Pull request by taeguk**
Check Box
---------------
- [x] Implementation of parallel::inplace_merge only for random access iterator.
- [x] Unit tests.
- [x] Benchmark codes.
- [x] Adapt to Ranges TS.

Issues
-------------
1. How to support bidirectional iterator. (#2826)
   -> For now, restrict the requirements of parallel::inplace_merge to only random access iterator.

Note for future
---------------------
1. parallel::inplace_merge is about 2 times or more faster than sequential thing.
2. Utilizing 2 cores or 4 cores is the limitation of the algorithm. More cores can't improve performance. I assume performance bottleneck is due to memory bandwidth.

### CODE REVIEW

**Comment by hkaiser** [comment]
@taeguk Could you add the new file to the build system, please (see here: https://github.com/STEllAR-GROUP/hpx/blob/master/docs/CMakeLists.txt#L46-L196)? Also, could you reference the new algorithm for the documentation index as well (here: https://github.com/STEllAR-GROUP/hpx/blob/master/docs/hpx.idx#L165-L348)?

File: hpx/parallel/algorithms/merge.hpp [hunk 1]
```diff
@@ -491,6 +490,341 @@ namespace hpx { namespace parallel { inline namespace v1
...
+            std::inplace_merge(first, middle, last,
+                util::compare_projected<Comp, Proj>(
+                    std::forward<Comp>(comp),
+                    std::forward<Proj>(proj)));
+            return last;
+        }
+
+        template <typename ExPolicy, typename RandIter,
+            typename Comp, typename Proj,
+        HPX_CONCEPT_REQUIRES_(
```
Comment (hkaiser) [inline hpx/parallel/algorithms/merge.hpp]: Do we have to use concepts for this internal API? What is your rationale of using it here?

File: hpx/parallel/algorithms/merge.hpp [hunk 2]
```diff
@@ -491,6 +490,341 @@ namespace hpx { namespace parallel { inline namespace v1
...
+            parallel(ExPolicy && policy,
+                RandIter first, RandIter middle, RandIter last,
+                Comp && comp, Proj && proj)
+            {
+                typedef util::detail::algorithm_result<
+                    ExPolicy, RandIter
+                > algorithm_result;
+
+                try {
+                    return algorithm_result::get(
```
Comment (hkaiser) [inline hpx/parallel/algorithms/merge.hpp]: If implementing a asynchronous version of `inplace_merge` is not feasible/necessary, could you at least run the algorithm on a new thread for `par(task)` and similar?

File: hpx/parallel/algorithms/merge.hpp (hunk 1 above)
Comment (taeguk) [reply hpx/parallel/algorithms/merge.hpp]: @hkaiser I just wanted to restrict that ```RandIter``` is only random access iterator.
Orginally, inplace_merge have to support bidirectional iterator, too.
But, in now, it only supports random access iterator.
I thought that I might have to implement different helper function for the bidirectional iterator in the future.
In that context, I used concepts for this helper function.

If it is strange and misuse of concepts, I will modify this. Is this misuse of concepts?

File: hpx/parallel/algorithms/merge.hpp [hunk 3]
```diff
@@ -491,6 +490,341 @@ namespace hpx { namespace parallel { inline namespace v1
...
+                }
+
+                fut.get();
+            }
+        }
+
+        template <typename ExPolicy, typename RandIter,
+            typename Comp, typename Proj,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::traits::is_random_access_iterator<RandIter>::value)>
```
Comment (taeguk) [inline hpx/parallel/algorithms/merge.hpp]: @hkaiser Is this use of concepts strange, too?

File: hpx/parallel/algorithms/merge.hpp (hunk 2 above)
Comment (hkaiser) [reply hpx/parallel/algorithms/merge.hpp]: I was just wondering. No need to change things.

File: hpx/parallel/algorithms/merge.hpp (hunk 3 above)
Comment (hkaiser) [reply hpx/parallel/algorithms/merge.hpp]: No need to change things.

File: hpx/parallel/algorithms/merge.hpp (hunk 2 above)
Comment (taeguk) [reply hpx/parallel/algorithms/merge.hpp]: @hkaiser Oh.. I was very fool.... :(
I don't know why I didn't consider asynchronous execution policy..
As you say, I modify the algorithm will run on a new thread.
There are similar problems in parallel::merge and parallel::partition, too.
I will fix them in another PR soon.

Comment (taeguk) [reply hpx/parallel/algorithms/merge.hpp]: @hkaiser The problem you suggested should be fixed as I think. Why I don't need to change things?

File: hpx/parallel/algorithms/merge.hpp (hunk 3 above)
Comment (taeguk) [reply hpx/parallel/algorithms/merge.hpp]: @hkaiser But, I removed that concept for consistency of codes.

File: hpx/parallel/algorithms/merge.hpp (hunk 2 above)
Comment (hkaiser) [reply hpx/parallel/algorithms/merge.hpp]: What I meant is that there is no need to remove the concept checks. 
I'd very much appreciate you adding the asynchronous behavior, though.

**Comment by taeguk** [comment]
@hkaiser I referenced the new algorithm for the documentation index as you said.
And I also resolved merge conflicts.

**Comment by hkaiser** [comment]
@taeguk do you plan do add the async implementations to this PR or would that be another one? IOW, can we go ahead with merging this PR now?

**Comment by taeguk** [comment]
@hkaiser I'm ready to be merged. In now, I supported async with ```execution::async_execute(..)```.
Do you want different async implementation for ```inplace_merge```?

**Pull request merged**